ambry-store/src/main/java/com.github.ambry.store/CompactionManager.java

```diff
@@ -52,4 +110,82 @@ CompactionDetails getCompactionDetails(BlobStore blobStore) throws StoreExceptio
     }
     return details;
   }
+
+  /**
+   * A {@link Runnable} that cycles through the stores and executes compaction if required.
+   */
+  private class CompactionExecutor implements Runnable {
+    private final Logger logger = LoggerFactory.getLogger(getClass());
+    private final AtomicBoolean enabled = new AtomicBoolean(true);
+    private final ReentrantLock lock = new ReentrantLock();
+    private final Condition waitCondition = lock.newCondition();
+    private final Set<BlobStore> storesToSkip = new HashSet<>();
+    private final long waitTimeMs =
+        storeConfig.storeCompactionCheckFrequencyInHours * Time.MinsPerHour * Time.SecsPerMin * Time.MsPerSec;
+
+    /**
+     * Starts by resuming any compactions that were left halfway. In steady state, it cycles through the stores at a
+     * configurable frequency and runs compaction as required.
+     */
+    @Override
+    public void run() {
+      // complete any compactions in progress
+      for (BlobStore store : stores) {
+        if (store.isStarted()) {
+          try {
+            store.maybeResumeCompaction();
+          } catch (Exception e) {
+            logger.error("Compaction of store {} failed on resume. Continuing with the next store", store, e);
+            storesToSkip.add(store);
+          }
+        }
+      }
+      // continue to do compactions as required.
+      while (enabled.get()) {
+        try {
+          long startTimeMs = time.milliseconds();
+          for (BlobStore store : stores) {
+            try {
+              if (!enabled.get()) {
+                break;
+              }
+              if (store.isStarted() && !storesToSkip.contains(store)) {
+                CompactionDetails details = getCompactionDetails(store);
+                if (details != null) {
+                  store.compact(details);
+                }
+              }
+            } catch (Exception e) {
+              logger.error("Compaction of store {} failed. Continuing with the next store", store, e);
+              storesToSkip.add(store);
+            }
+          }
+          if (enabled.get()) {
+            long timeElapsed = time.milliseconds() - startTimeMs;
```

Comment:
there could be a race condition here I guess. I would suggest to take the lock and then check the value of "enabled". Similarly in line 182, can we do if (enabled.compareAndSet(false, true)) { ...}. Also do take the lock before checking for "enabled" value.